Task 3

enginire4ek opened the pull request:
- User gets access to the personal account
- Some fixes in ProfileFragment.kt
- Added new pop-up in profile
- Added registration for user
- Added RegistrationFragment
- Added SessionManager that works with SharedPreferences
- Added two new requests in ApiService.kt
- A request body with a cookie is attached to requests
- TODO: authorization

enginire4ek committed:
Task 3
- User gets access to the personal account
- Some fixes in ProfileFragment.kt
- Added new pop-up in profile
- Added registration for user
- Added RegistrationFragment
- Added SessionManager that works with SharedPreferences
- Added two new requests in ApiService.kt
- A request body with a cookie is attached to requests
- TODO: authorization

102Rudy commented on app/src/main/java/com/example/pinterbest/ProfileFragment.kt:
```diff
@@ -19,6 +35,12 @@ class ProfileFragment : Fragment() {
+    var followers: Int? = null
+    var following: Int? = null
+
+    private val apiService = ApiClient().retrofit.create(ApiService::class.java)
```
you are creating ApiService instance in few places. Let's make it singleton

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/f47063b96f0a24a56ad6def81896bbe4dee63167

102Rudy commented on app/src/main/java/com/example/pinterbest/data/repository/Repository.kt:
```diff
@@ -45,7 +47,7 @@ class Repository(val apiService: ApiService) {
 
-    fun getpinFeed() = result {
-        apiService.getPinFeed()
+    fun getpinFeed(context: Context) = result {
```
context in repositories is prohibited. Repository must know nothing about Android framework

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/f47063b96f0a24a56ad6def81896bbe4dee63167

102Rudy commented on app/src/main/java/com/example/pinterbest/utilities/SessionManager.kt:
```diff
@@ -0,0 +1,44 @@
+import android.content.SharedPreferences
+import com.example.pinterbest.R
+
+class SessionManager(context: Context) {
```
The manager is kind of a god object. Consider break down this class into known components, such as Interactor, Repository

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/f47063b96f0a24a56ad6def81896bbe4dee63167

102Rudy commented on app/src/main/java/com/example/pinterbest/utilities/Validator.kt:
```diff
@@ -1,10 +1,21 @@
 object Validator {
     private const val PASSWORD_MIN_SIZE = 8
+    private val PASSWORD_POLICY =
+        """Пароль должен состоять минимум из 8 символов, а также
+            |- должен содержать хотя бы одну заглавную букву,
+            |- хотя бы одну строчную букву,
+            |- хотя бы одну цифру и
+            |- хотя бы один специальный символ среди ~!@#$%^&*()-_=+|[]{};:'\",<.>/?""".trimMargin()
+
+    private const val NAME_VALIDATION_MSG = "Введите корректное имя"
+    private const val EMAIL_VALIDATION_MSG = "Введите корректный электронный адрес"
 
```
nit: let's put it into android resources

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/f47063b96f0a24a56ad6def81896bbe4dee63167

102Rudy commented on app/src/main/java/com/example/pinterbest/viewmodels/HomeViewModel.kt:
```diff
@@ -1,10 +1,11 @@
 import com.example.pinterbest.data.repository.Repository
 
-class HomeViewModel(repository: Repository) :
+class HomeViewModel(context: Context, repository: Repository) :
```
ViewModels should not to know about context, it's better to keep them without it to make them testable and cleaner

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/f47063b96f0a24a56ad6def81896bbe4dee63167

enginire4ek committed:
Task 3. Updates
- Added thread-safe creation of ApiClient instance
- Added two functions for working with user authorization in ApiService.kt
- Have been created new classes of models and factories. In the future, using code generation, the code will be simplified
- Now user can interact with Login screen, Home screen and Profile screen
- Renamed repository for working with session SessionRepository.kt

102Rudy commented on app/src/main/java/com/example/pinterbest/LoginFragment.kt:
```diff
@@ -1,19 +1,31 @@
     private var _binding: FragmentLoginBinding? = null
     private val binding get() = _binding!!
 
+    private lateinit var sessionManager: SessionRepository
```
sessionRepository

enginire4ek replied:
Renamed

102Rudy commented on app/src/main/java/com/example/pinterbest/LoginFragment.kt:
```diff
@@ -24,30 +36,127 @@ class LoginFragment : Fragment() {
+    }
+
+    private fun saveSession(response: Response<ResponseBody>) {
+        val regex =
+            """['session_id=']+(?<=session_id=).{56}""".toRegex()
+        val cookie = regex.find(response.headers().values("Set-Cookie")[0])
+        sessionManager.saveUserData(cookie?.value ?: "")
```
UI layer is bad place for this logic. You need interactor for it. Also expressions with regex look weird, why don't you just access `Set-Cookie` header?
In case there is no way to implement it without this I suggest to consider some CookieProvider with the following contract

```kotlin
interface CookieProvider {
    // returns cookie string
    fun getCookie(response: Response<*>): String
}
```

enginire4ek replied:
Created AuthRepository

102Rudy commented on app/src/main/java/com/example/pinterbest/LoginFragment.kt:
```diff
@@ -24,30 +36,127 @@ class LoginFragment : Fragment() {
+            .getItem(MainActivity.HOME_POSITION_BNV).isChecked = true
+    }
+
+    private fun loginProvider(view: View, response: Response<ResponseBody>) {
```
actually, `Response` is a data layer element. It shouldn't be used on UI level
I suggest you to operate with exceptions to implement the following logic. But mapping response.code to exception or data move to Repository (data layer)

enginire4ek replied:
Created AuthRepository

102Rudy commented on app/src/main/java/com/example/pinterbest/LoginFragment.kt:
```diff
@@ -24,30 +36,127 @@ class LoginFragment : Fragment() {
+    }
+
+    companion object {
+        const val SUCCESS = 204
+        const val INVALID_DATA = 400
+        const val WRONG_PASSWORD = 401
+        const val ALREADY_AUTHORIZED = 403
+        const val USER_NOT_FOUND = 404
```
this codes are also data layer entities as well as `Response`

enginire4ek replied:
Yes

102Rudy commented on app/src/main/java/com/example/pinterbest/api/ApiClient.kt:
```diff
@@ -9,6 +9,31 @@ import retrofit2.Retrofit
+    private var mRetrofit: Retrofit
+
+    init {
+        mRetrofit = Retrofit.Builder()
```
you can avoid `init` and make mRetrofit va**L** if rewrite:
```kotlin
private val mRetrofit: Retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
             .addConverterFactory(GsonConverterFactory.create())
             .client(okHttpClient().build())
             .build()
```

enginire4ek replied:
Fixed here https://github.com/enginire4ek/Pinterbest/pull/3/commits/5fc6193ab1b0ed6d7b799a8929f6e95603658369

enginire4ek committed:
Task 3. Updates
- Changed the logic for obtaining a token
- Creating AuthRepository in which the data is processed

enginire4ek committed:
Task 3. Updates
- Architectural changes in authorization and registration

enginire4ek committed:
Task 3. Updates
- Small fixes in navigation

enginire4ek committed:
Task 3. Updates
- authProvider in AuthRepository.kt throws custom Exceptions when the response code is bad
- Added application action when there is no Internet connection. Now a fragment_error.xml is shown, which indicates the lack of an Internet connection
- Mutated from MutableLiveData to SingleLiveEvent
- Changed LoginViewModel.kt and RegistrationViewModel.kt
- Changed LoginFactory.kt and RegistrationFactory.kt

enginire4ek committed:
Task 3. Updates
- Added nice animated placeholder on Profile screen

102Rudy commented on app/src/main/java/com/example/pinterbest/LoginFragment.kt:
```diff
@@ -24,19 +41,94 @@ class LoginFragment : Fragment() {
+                try {
+                    AuthRepository(sessionRepository).authProvider(response)
+                    view.findNavController().navigate(R.id.homeFragment)
+                    setUpBottomNavigationItem()
+                } catch (t: IllegalStateException) {
+                    showErrorToast(t)
+                }
+            }
+        )
+    }
+
+    private fun <T> showErrorToast(t: T) {
+        when (t) {
+            is AlreadyAuthorizedException -> {
+                Toast.makeText(
+                    context,
+                    resources.getString(R.string.error_already_authorized),
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+            is InvalidDataException -> {
+                Toast.makeText(
+                    context,
+                    resources.getString(R.string.error_invalid_data),
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+            is WrongPasswordException -> {
+                Toast.makeText(
+                    context,
+                    resources.getString(R.string.error_wrong_password),
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+            is UserNotFoundException -> {
+                Toast.makeText(
+                    context,
+                    resources.getString(R.string.error_user_not_found),
+                    Toast.LENGTH_SHORT
+                ).show()
```
wdyt about moving part of this logic into ViewModel? e.g. mapping from exception to string res id?
It's better to keep fragments as simple as possible, without non presentation logic

102Rudy commented on app/src/main/java/com/example/pinterbest/data/repository/AuthRepository.kt:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    private fun throwInvalidDataException() {
+        throw InvalidDataException()
+    }
+
+    private fun throwWrongPasswordException() {
+        throw WrongPasswordException()
+    }
+
+    private fun throwAlreadyAuthorizedException() {
+        throw AlreadyAuthorizedException()
+    }
+
+    private fun throwUserNotFoundException() {
+        throw UserNotFoundException()
+    }
+
+    private fun throwUserExistsException() {
+        throw UserExistsException()
+    }
```
you don't need this code. Just throw exceptions where needed, without special methods

enginire4ek replied:
https://detekt.github.io/detekt/style.html#throwscount

enginire4ek committed:
Task 3. Updates
- Removed unnecessary logic from fragments
- Changed max ThrowsCount to 6

enginire4ek committed:
Task 3. Updates
- Error handling moved to a separate place
